### DISCUSSION TITLE
variable substitution with MQTT

### DISCUSSION THREAD

**Discussion by hallard** [arendst/Tasmota#10933] [Q&A] [answered]
Would it be possible to have internal variable substitution on MQTT full topic, such as for example:

```
emoncms/%topic%/%prefix%/%var1%/
```

**Comment by barbudor** [comment, marked as answer]
Hi Charles,

Could you please elaborate on this ?
What is the use case ?

It doesn't make sense to have the device topic to depend on a variable because Tasmota subscribe to command topic at startup and do not support dynamically changing the topic. As of today, changing Topic will restart Tasmota.

However in rules you can use %topic%. For example you can use `publish emoncs/%topic%/stat/%var1% somedata`

**Reply by hallard** [reply]
Sure,

The idea behind this is the ability to store an API key, when the device publish to MQTT and if after we need to publish data to another platform (emoncms in my case) I need an API key and I would like to avoid storing a mapping between device and api key on node red. From my point of view API key should be in the device itself. The idea is to put the API key into the topic name to be able to extract it on the other side to send data afterwards to emoncms (my dream would be adding tasmota http post feature but that is not his main goal I know)

`/emoncms/denky_xxyyzz/apikey/tele/#`

Not sure it's the best idea, but since reading you answer, I can use `%topic%` and some others in rules I would be able to deal with that. I found the best idea for now is to set the API key in field `client_id`of MQTT config but unfortunately node red is unable to return back the client id. Is there any solution such as on boot:

 `publish emoncs/%topic%/stat/apikey %cliend_name_mqtt_confid%` retained

or every 5 minutes for example resend the message `tele/INFO1` which contain the mqtt client name field in `FallbackTopic` (but I don't know how to do this one even if it's not the cleanest on my point of view)

![image](https://user-images.githubusercontent.com/2471931/107846470-54f14080-6de4-11eb-9fb7-82fa0f1cdb32.png)

Any

**Reply by barbudor** [reply]
At some point you need to configure the API key on the device.
So why not directly entering it in the full topic ?
Remember that `fulltopic` can support already some variable substitutions :
>  Use of optional %prefix%, %topic%, %hostname%, and %id% substitution tokens is allowed.

So your idea of using the MqttClient is a good idea but instead of trying to get it from the MQTT connection, use it into the fulltopic:
`fulltopic emoncms/%topic%/%prefix/%id%`
should achieve what you are looking for.
But don't forget that not only your state and tele messages will include the API key but the command topic also will require the API key so you will need to use `emoncms/%topic%/cmnd/APIKey/anycommand` to send anycommand to the device.

If you want this to happen only on published status messages, you have to cook your own publish with rules where you can use a %var% or a %mem% substitution.

**Reply by hallard** [reply]
@barbudor 
Thanks for your quick answer
Sure, put the APIkey on topic is an option but not for large deployment where I need to keep basic config common to all devices (and then just need to set api key somewhere). 

Yes I already looked at the `%id%` field of the topic but it's the MAC Address of the device not the `MQTT client name`, of course the `MQTT client name` would have saved my life, it's the 1st thing I looked for.

While writing may be the best option would have a new MQTT field called  `%client_id%` or `%client_name%`, makes sense :-)

**Reply by barbudor** [reply]
You are correct.
I was foolled because my MqttClient is my MAC address (MqttClient "%12X") :(
